services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java

```diff
@@ -673,10 +677,7 @@ protected AccessToken initToken(RealmModel realm, ClientModel client, UserModel
         token.setNonce(clientSessionCtx.getAttribute(OIDCLoginProtocol.NONCE_PARAM, String.class));
         token.setScope(clientSessionCtx.getScopeString());
 
-        // Best effort for "acr" value. Use 0 if clientSession was authenticated through cookie ( SSO )
-        // TODO: Add better acr support. See KEYCLOAK-3314
-        String acr = (AuthenticationManager.isSSOAuthentication(clientSession)) ? "0" : "1";
-        token.setAcr(acr);
+        token.setAcr(getAcr(clientSession));
```

Comment:
Wouldn't be more natural to map `acr` using a protocol mapper? Possibly having a default client scope for this purpose.

Reply:
Shall we set the `acr` only if requested either by `claims` or `acr_values`?

Reply:
We implemented that according to the already existing code. If you want to do it as a protocol mapper, that's ok for us.
The same applies to the question whether to set the acr value only when requested or not. This could also be made configurable in the protocol mapper.

Reply:
Sounds good to me.